silq/analysis/analysis.py

```diff
@@ -715,8 +717,81 @@ def analyse_EPR(empty_traces: np.ndarray,
+
+    Using kernel density estimation, 'determine_threshold_up_proportion' function
+    determines the probability density distribution of measured electron spin-up
+    proportions within the 0 to 1 up proportion space in the form of a Gaussian function.
+    It is later analysed for the peaks in the density distribution and based on their
+    relative position within 0 to 1 up proportion space, the density minimum is found,
+    which is considered to be the threshold for determining the nuclear spin state.
+
+    Args:
+        up_proportions_arr: 1D Array of up proportions, calculated from
+            `analyse_traces`. Up proportion is the proportion of traces that
+            contain blips. The length is number of samples.
+        shots_per_frequency: Integer number of ESR shots/traces, which are
+            used to determine the up proportion.
+
+    Returns:
+        (float) threshold up-proportion
+    """
+
+    num_shots = max(shots_per_frequency, 25)
+    # Algorithm fails if the number of shots is less than 25 (not exact number, can still be adjusted).
+    # The reason for this is that in the following division of up_proportion space by the number of shots
+    # there are too few points to perform kernel density estimation.
+
+    proportion_space = np.linspace(0, 1, num=num_shots + 1, endpoint=True)
+    up_proportions_arr = up_proportions_arr.reshape(1, -1)    # don't think it's needed
+    kernel = gaussian_kde(up_proportions_arr)
+    gaussian_up_proportions = kernel(proportion_space)
+
+    # Finding indexes of density distribution peaks. Might still require some fine adjustments in
+    # 'thres' and 'min_dist' parameters.
+    peak_idxs = peakutils.peak.indexes(gaussian_up_proportions,
+                                       thres=0.5/up_proportions_arr.shape[-1],
+                                       min_dist=num_shots/5)
+    # If no peaks are detected (something failed), by default we set threshold to 0.5.
+    middle_point = (num_shots + 1) // 2
+    if len(peak_idxs) == 0:
+        logger.debug(f'Adaptive thresholding routine: 0 peaks were '
+                     f'found, using threshold 0.5')
+        trough_idx = middle_point
+    # If only one peak is detected (nuclear spin did not flip), depending on whether
+    # it is in the lower (<0.5) or the upper (>=0.5) half of proportion space,
+    # the minimum index of density distribution is found respectively above or below
+    # this peak's index.
+    # Since the density distribution values reach 1e-20 at the edges of the
+    # proportion space (i.e. at 0 or 1), to smooth a bit the threshold jumps between
+    # the measurement points, rounding to the 4th digit is implemented.
+    elif len(peak_idxs) == 1:
+        if peak_idxs[0] < middle_point:
+            trough_slice = slice(*[peak_idxs[0], num_shots + 1])
+            search_space = np.round(gaussian_up_proportions[trough_slice], 4)
+            trough_idx = peak_idxs[0] + np.argmin(search_space)
```

Comment:
I'm not following these lines, it needs a bit more commenting. Are you basically finding the first up proportion above the peak and setting that as your threshold? I'm not seeing why that is the right approach, can you explain?

Reply:
No, if there is only one peak, it finds its index, then it looks for the minimum of density distribution (somewhere far from the Gaussian peak) in the proportion space between the peak maximum and the rest of proportion space. There is a condition that it searches for density minimum in proportion space that is larger. E.g. if peak is at 0.2 proportion space, minimum will be found in the range between 0.2 and 1. If peak was at 0.6, the minimum would be found somewhere between 0 and 0.6.

Reply:
I added these comments below. Should I add more?
# If only one peak is detected (nuclear spin did not flip), depending on whether
    # it is in the lower (<0.5) or the upper (>=0.5) half of proportion space,
    # the minimum index of density distribution is found respectively above or below
    # this peak's index.
    # Since the density distribution values reach 1e-20 at the edges of the
    # proportion space (i.e. at 0 or 1), to smooth a bit the threshold jumps between
    # the measurement points, rounding to the 4th digit is implemented.

Reply:
yep I think it's good